core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -118,9 +119,18 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
             ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, nodeEnv.availableShardPaths(request.shardId));
             if (shardStateMetaData != null) {
                 final IndexMetaData metaData = clusterService.state().metaData().index(shardId.index().name()); // it's a mystery why this is sometimes null
-                if (metaData != null && canOpenIndex(request.getShardId(), metaData) == false) {
-                    logger.trace("{} can't open index for shard [{}]", shardId, shardStateMetaData);
-                    return new NodeGatewayStartedShards(clusterService.localNode(), -1);
+                if (metaData != null) {
+                    ShardPath shardPath = null;
+                    try {
+                        shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, metaData.settings());
+                        if (shardPath == null) {
+                            throw new IllegalStateException(shardId + " no shard state found");
+                        }
+                        Store.tryOpenIndex(shardPath.resolveIndex());
+                    } catch (Exception exception) {
+                        logger.trace("{} can't open index for shard [{}] in path [{}]", exception, shardId, shardStateMetaData, (shardPath != null) ? shardPath.resolveIndex() : "");
+                        return new NodeGatewayStartedShards(clusterService.localNode(), -1, exception);
```

Comment:
now we report the correct version even in case of existing store exception. `GatewayAllocator` treats shards with store exception the same way as it does non-existent shards on node.